File: frontend/src/lib/schemas/momentum.ts
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Sub-indicator schemas
// ---------------------------------------------------------------------------

export const rsiIndicatorSchema = z.object({
  value: z.number().nullable(),
  /** Raw 0-100 score before weight is applied */
  raw_score: z.number().int().min(0).max(100),
  /** Weighted contribution: raw_score × 0.20, capped at 20 */
  score: z.number().int().min(0).max(20),
  max_score: z.number().int().default(20),
});

export const macdIndicatorSchema = z.object({
  macd_line: z.number(),
  signal_line: z.number(),
  histogram: z.number(),
  /** Raw 0-100 score before weight is applied */
  raw_score: z.number().int().min(0).max(100),
  /** Weighted contribution: raw_score × 0.15, capped at 15 */
  score: z.number().int().min(0).max(15),
  max_score: z.number().int().default(15),
});

export const maAlignmentIndicatorSchema = z.object({
  ma_20: z.number(),
  ma_50: z.number(),
  ma_200: z.number(),
  /** ABOVE_ALL | ABOVE_50_200 | ABOVE_200 | BELOW_ALL | INSUFFICIENT_DATA */
  label: z.string(),
  /** Raw 0-100 score before weight is applied */
  raw_score: z.number().int().min(0).max(100),
  /** Weighted contribution: raw_score × 0.20, capped at 20 */
  score: z.number().int().min(0).max(20),
  max_score: z.number().int().default(20),
});

export const week52PositionIndicatorSchema = z.object({
  high_52w: z.number(),
  low_52w: z.number(),
  /** 0 = at 52w low, 100 = at 52w high */
  position_pct: z.number().min(0).max(100),
  /** Raw 0-100 score before weight is applied */
  raw_score: z.number().int().min(0).max(100),
  /** Weighted contribution: raw_score × 0.15, capped at 15 */
  score: z.number().int().min(0).max(15),
  max_score: z.number().int().default(15),
});

export const performanceIndicatorSchema = z.object({
  perf_1m: z.number(),
  perf_6m: z.number(),
  /** Raw 0-100 score for 1-month return before weight */
  raw_score_1m: z.number().int().min(0).max(100),
  /** Raw 0-100 score for 6-month return before weight */
  raw_score_6m: z.number().int().min(0).max(100),
  /** 1-month weighted contribution: raw × 0.15, capped at 15 */
  score_1m: z.number().int().min(0).max(15),
  /** 6-month weighted contribution: raw × 0.10, capped at 10 */
  score_6m: z.number().int().min(0).max(10),
  score: z.number().int().min(0).max(25),
  max_score: z.number().int().default(25),
});

export const sectorMomentumIndicatorSchema = z.object({
  sector_etf: z.string(),
  ticker_perf_6m: z.number(),
  sector_perf_6m: z.number(),
  relative_perf_6m: z.number(),
  /** Raw 0-100 score before weight is applied */
  raw_score: z.number().int().min(0).max(100),
  /** Weighted contribution: raw_score × 0.05, capped at 5 */
  score: z.number().int().min(0).max(5),
  max_score: z.number().int().default(5),
});

// ---------------------------------------------------------------------------
// Top-level response schema
// ---------------------------------------------------------------------------

export const momentumResponseSchema = z.object({
  ticker: z.string(),
  sector_etf: z.string(),
  rsi: rsiIndicatorSchema,
  macd: macdIndicatorSchema,
  ma_alignment: maAlignmentIndicatorSchema,
  week_52_position: week52PositionIndicatorSchema,
  performance: performanceIndicatorSchema,
  sector_momentum: sectorMomentumIndicatorSchema,
  /** 0–100 composite F1 score */
  f1_score: z.number().int().min(0).max(100),
  /** STRONG BUY | BUY | NEUTRAL | WEAK | AVOID */
  f1_grade: z.string(),
});

// ---------------------------------------------------------------------------
// Exported TypeScript types
// ---------------------------------------------------------------------------

export type RsiIndicator = z.infer<typeof rsiIndicatorSchema>;
export type MacdIndicator = z.infer<typeof macdIndicatorSchema>;
export type MaAlignmentIndicator = z.infer<typeof maAlignmentIndicatorSchema>;
export type Week52PositionIndicator = z.infer<typeof week52PositionIndicatorSchema>;
export type PerformanceIndicator = z.infer<typeof performanceIndicatorSchema>;
export type SectorMomentumIndicator = z.infer<typeof sectorMomentumIndicatorSchema>;
export type MomentumResponse = z.infer<typeof momentumResponseSchema>;

File: backend/src/atlas/schemas/momentum.py
"""Pydantic schemas for the F1 Momentum endpoint."""

from __future__ import annotations

from typing import Final

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# F1 grade literals
# ---------------------------------------------------------------------------


class F1Grade:
    """Named constants for the five F1 momentum grades."""

    STRONG_BUY: Final[str] = "STRONG BUY"
    BUY: Final[str] = "BUY"
    NEUTRAL: Final[str] = "NEUTRAL"
    WEAK: Final[str] = "WEAK"
    AVOID: Final[str] = "AVOID"


# ---------------------------------------------------------------------------
# Sub-indicator schemas
# ---------------------------------------------------------------------------


class RsiIndicator(BaseModel):
    """RSI (14-period Wilder) result and its contribution to the F1 score."""

    model_config = ConfigDict(from_attributes=True)

    value: float | None = Field(
        None, description="RSI value in [0, 100]; null when insufficient data."
    )
    raw_score: int = Field(
        ge=0, le=100, description="RSI raw score (0-100) before weight is applied."
    )
    score: int = Field(
        ge=0, le=20, description="Weighted contribution to F1 (raw_score x 0.20, 0-20)."
    )
    max_score: int = Field(
        default=20, description="Maximum contribution this indicator can add to F1."
    )


class MacdIndicator(BaseModel):
    """MACD (12/26/9) result and its contribution to the F1 score."""

    model_config = ConfigDict(from_attributes=True)

    macd_line: float = Field(description="MACD line value (fast EMA - slow EMA).")
    signal_line: float = Field(description="Signal line (9-period EMA of MACD).")
    histogram: float = Field(description="MACD histogram (macd_line - signal_line).")
    raw_score: int = Field(
        ge=0, le=100, description="MACD raw score (0-100) before weight is applied."
    )
    score: int = Field(
        ge=0, le=15, description="Weighted contribution to F1 (raw_score x 0.15, 0-15)."
    )
    max_score: int = Field(default=15)


class MaAlignmentIndicator(BaseModel):
    """20/50/200-day MA alignment result and its contribution to the F1 score."""

    model_config = ConfigDict(from_attributes=True)

    ma_20: float = Field(description="20-day simple moving average.")
    ma_50: float = Field(description="50-day simple moving average.")
    ma_200: float = Field(description="200-day simple moving average.")
    label: str = Field(
        description=(
            "Alignment label: ABOVE_ALL | ABOVE_50_200 | ABOVE_200 | BELOW_ALL | INSUFFICIENT_DATA"
        )
    )
    raw_score: int = Field(
        ge=0, le=100, description="MA alignment raw score (0-100) before weight is applied."
    )
    score: int = Field(
        ge=0, le=20, description="Weighted contribution to F1 (raw_score x 0.20, 0-20)."
    )
    max_score: int = Field(default=20)


class Week52PositionIndicator(BaseModel):
    """52-week high/low position result and its contribution to the F1 score."""

    model_config = ConfigDict(from_attributes=True)

    high_52w: float = Field(description="52-week closing high.")
    low_52w: float = Field(description="52-week closing low.")
    position_pct: float = Field(
        ge=0.0,
        le=100.0,
        description="Where the current price sits in the 52-week range (0=low, 100=high).",
    )
    raw_score: int = Field(
        ge=0, le=100, description="52-week raw score (0-100) before weight is applied."
    )
    score: int = Field(
        ge=0, le=15, description="Weighted contribution to F1 (raw_score x 0.15, 0-15)."
    )
    max_score: int = Field(default=15)


class PerformanceIndicator(BaseModel):
    """1-month and 6-month price performance and their combined contribution."""

    model_config = ConfigDict(from_attributes=True)

    perf_1m: float = Field(description="1-month (21-session) total return (%).")
    perf_6m: float = Field(description="6-month (126-session) total return (%).")
    raw_score_1m: int = Field(
        ge=0, le=100, description="1-month raw score (0-100) before weight is applied."
    )
    raw_score_6m: int = Field(
        ge=0, le=100, description="6-month raw score (0-100) before weight is applied."
    )
    score_1m: int = Field(
        ge=0, le=15, description="1-month weighted contribution (raw x 0.15, 0-15)."
    )
    score_6m: int = Field(
        ge=0, le=10, description="6-month weighted contribution (raw x 0.10, 0-10)."
    )
    score: int = Field(
        ge=0, le=25, description="Combined performance contribution (score_1m + score_6m, 0-25)."
    )
    max_score: int = Field(default=25)


class SectorMomentumIndicator(BaseModel):
    """Sector relative performance result and its contribution to the F1 score."""

    model_config = ConfigDict(from_attributes=True)

    sector_etf: str = Field(description="SPDR sector ETF used as the benchmark (e.g. XLK).")
    ticker_perf_6m: float = Field(description="Ticker 6-month total return (%).")
    sector_perf_6m: float = Field(description="Sector ETF 6-month total return (%).")
    relative_perf_6m: float = Field(description="Ticker minus sector 6-month return (%).")
    raw_score: int = Field(
        ge=0, le=100, description="Sector raw score (0-100) before weight is applied."
    )
    score: int = Field(
        ge=0, le=5, description="Weighted contribution to F1 (raw_score x 0.05, 0-5)."
    )
    max_score: int = Field(default=5)


# ---------------------------------------------------------------------------
# Top-level response schema
# ---------------------------------------------------------------------------


class MomentumResponse(BaseModel):
    """Complete F1 Momentum analysis for a single ticker."""

    model_config = ConfigDict(from_attributes=True)

    ticker: str = Field(description="Ticker symbol (upper-case).")
    sector_etf: str = Field(description="Sector ETF used for relative momentum.")
    rsi: RsiIndicator
    macd: MacdIndicator
    ma_alignment: MaAlignmentIndicator
    week_52_position: Week52PositionIndicator
    performance: PerformanceIndicator
    sector_momentum: SectorMomentumIndicator
    f1_score: int = Field(ge=0, le=100, description="Composite F1 Momentum score (0-100).")
    f1_grade: str = Field(
        description="F1 grade: STRONG BUY | BUY | NEUTRAL | WEAK | AVOID"
    )
